### ISSUE TITLE
NaN gradients even after removing 0/0 value using torch.where

### ISSUE THREAD

**Issue by hakanardo** [pytorch/pytorch#16317]
## 🐛 Bug

If I use torch.where to replace NaN values in a tensor with a constant, the gradinets still become NaN.

## To Reproduce

The following snipped:
```python
x = torch.tensor([0.0, 1.0], requires_grad=True)
n = torch.tensor([0.0, 2.0])
y = x / n
y = torch.where(n==0, torch.zeros_like(y), y)
y.sum().backward()
print(y)
print(x.grad)
```

Results in
```
tensor([0.0000, 0.5000], grad_fn=<SWhereBackward>)
tensor([   nan, 0.5000])
```

## Expected behavior
The gradient should be 0:
```
tensor([0.0000, 0.5000], grad_fn=<SWhereBackward>)
tensor([0.0000, 0.5000])
```
to be consistent with other implementations of the same function such as for example:
```python
x = torch.tensor([0.0, 1.0], requires_grad=True)
n = torch.tensor([0.0, 2.0])
y = x / torch.where(n==0, torch.ones_like(n), n)
y = torch.where(n==0, torch.zeros_like(y), y)
y.sum().backward()
print(y)
print(x.grad)
```

## Environment

Please copy and paste the output from our
[environment collection script](https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py)
(or fill out the checklist below manually).

You can get the script and run it with:
```
wget https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py
# For security purposes, please check the contents of collect_env.py before running it.
python collect_env.py
```

PyTorch version: 1.0.0
Is debug build: No
CUDA used to build PyTorch: 9.0.176

OS: Debian GNU/Linux 9.4 (stretch)
GCC version: (Debian 6.3.0-18+deb9u1) 6.3.0 20170516
CMake version: version 3.7.2

Python version: 3.5
Is CUDA available: No
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA

Versions of relevant libraries:
[pip3] numpy (1.16.0)
[pip3] torch (1.0.0)
[pip3] torchtext (0.2.3)
[pip3] torchvision (0.2.1)
[conda] Could not collect

 - How you installed PyTorch (`conda`, `pip`, source): pip

**Comment by vishwakftw** [comment]
This is related to https://github.com/pytorch/pytorch/issues/15506.

**Comment by t-vi** [comment]
Indeed the first y would have gradient of 0. But then the div backward from ` / n` multiplies 0 with NaN giving NaN.

**Comment by zou3519** [comment]
Closing due to @t-vi's comment; due to how autograd works, even though you removed the NaN from a later part the NaNs still return because you have a step where you're dividing by 0.

**Comment by NagabhushanSN95** [comment]
But, how to solve this issue? Would `y = x / (n + 1e-3)` help?

**Comment by t-vi** [comment]
the easiest probably is to mask them in the backward.
